src/xform_to_datalog/generate_datalog_program.cc

```diff
@@ -0,0 +1,88 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+//
+// Tool that takes a manifest proto and generates corresponding datalog facts.
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/flags/usage.h"
+#include "src/common/logging/logging.h"
+#include "src/xform_to_datalog/datalog_facts.h"
+#include "src/xform_to_datalog/manifest_datalog_facts.h"
+
+ABSL_FLAG(std::string, datalog_file, "", "output file for the datalog facts");
+ABSL_FLAG(std::string, manifest_proto, "", "The manifest proto file.");
+ABSL_FLAG(bool, overwrite, false,
+          "Should we overwrite the output file if it exists.");
+
+constexpr char kUsageMessage[] =
+    "This tool takes a manifest proto and generates a datalog program.";
+
+int main(int argc, char* argv[]) {
+  google::InitGoogleLogging("generate_datalog_program");
+  absl::SetProgramUsageMessage(kUsageMessage);
+  absl::ParseCommandLine(argc, argv);
+
+  // Verify command line arguments
+  //
+  std::filesystem::path manifest_filepath(absl::GetFlag(FLAGS_manifest_proto));
+  if (!std::filesystem::exists(manifest_filepath)) {
+    LOG(ERROR) << "Manifest proto file " << manifest_filepath
+               << " does not exist!";
+    return -1;
+  }
+
+  bool force_overwrite = absl::GetFlag(FLAGS_overwrite);
+  std::filesystem::path datalog_filepath(absl::GetFlag(FLAGS_datalog_file));
+  if (std::filesystem::exists(datalog_filepath) && !force_overwrite) {
+    LOG(ERROR) << "Output datalog file " << datalog_filepath
+               << " exists! Forgot `--overwrite`?";
+    return -1;
+  }
+
+  // Parse manifest stream and dump as datalog program.
+  std::ifstream manifest_proto_stream(manifest_filepath);
+  if (!manifest_proto_stream) {
+    LOG(ERROR) << "Error reading manifest proto file " << manifest_filepath
+               << ":" << strerror(errno);
+    return -1;
+  }
+
+  arcs::ManifestProto manifest_proto;
+  if (!manifest_proto.ParseFromIstream(&manifest_proto_stream)) {
+    LOG(ERROR) << "Error parsing the manifest proto " << manifest_filepath;
+  }
+
+  auto datalog_facts = raksha::xform_to_datalog::DatalogFacts(
+      raksha::xform_to_datalog::ManifestDatalogFacts::CreateFromManifestProto(
+          manifest_proto));
+
+  std::ofstream datalog_file(
+      datalog_filepath, std::ios::out | std::ios::trunc | std::ios::binary);
```

Comment:
Why `ios::binary` here? Just curious.

Reply:
ios::binary makes sure newline characters are written as is. Without `\n` will be written as `\r\n` on Windows machines, for example.